### PR TITLE
Add thread-safe app handle

### PR DESCRIPTION
**Pull request by clavin**
### Problem

Currently, it's pretty rough to create a new window—or do any other operation Glazier requires happen on the main thread—from a non-main thread:

* If you have a window already, you can use an `IdleHandle` to schedule a callback on the main thread. This is alright, even though it requires the user to carry around the handle.
* If you _haven't_ created a window at all, then delay the `Application` from running until you're ready to create the first window.
* If you plan on creating a new window after destroying the last remaining one, then you must either:
  * Use a combination of both above workarounds, destroying the last remaining window on the main thread and then delaying it (which can cause the OS to believe your application has stopped responding!), waiting until you receive a signal to create a new window again.
  * Hide the last remaining window so when you need a new "initial" window you can (1) reuse it as-is, (2) reset it to its initial state (which isn't always very trivial), or (3) use it to access the main thread, create a new "initial" window, and destroy the old, hidden "last remaining" window.

After you call `Application::run` there is currently only one way to run code on the main thread[^1]: `IdleHandle::add_idle`. But the only way to have an `IdleHandle` is to have an existing window. If your application has no existing windows, you're left unable to run anything on the main thread. Thus, the user is forced to either emulate the state of no windows or freeze the main thread during the time in which there are no windows.

[^1]: The user may be given back control of the main thread after the `Application::run` loop terminates, but that's not very relevant here.

Put another way: access to the main thread revolves around _windows_, even though the encapsulating _application_ also exists only on the main thread. No windows = no access to main thread = main thread deadlock.

This issue pops up in any situation when there are no windows but your application is still running. Common examples of this are:

* macOS apps, where it is customary for apps to still be running without any windows
* System tray apps
* The transition between two isolated windows, for example from splash screen to initial window

As I highlighted upfront, there are workarounds to get this behavior somewhat working, however they end up making user code unnecessarily complex in practice.

### Proposed Solution

I propose adding an `AppHandle` struct that can be shared across threads, as well as an associated `Application:get_handle` method to create instances of that struct. Similar to `IdleHandle`, `AppHandle` simply allows the user to schedule a function to be run on the main thread (at some point or potentially never).

This creates an equivalent counterpart to `IdleHandle` for `Application`:

<table>
  <tr>
    <th>–</th>
    <th>App</th>
    <th>Window</th>
  </tr>
  <tr>
    <th><code>!Send</code>, <code>!Sync</code></th>
    <td><code>Application</code></td>
    <td><code>WindowHandle</code></td>
  </tr>
  <tr>
    <th><code>Send</code>, <del><code>Sync</code></del></th>
    <td><strong><code>AppHandle</code> ✨</strong></td>
    <td><code>IdleHandle</code></td>
  </tr>
</table>

> **Update**: The current implementation of `AppHandle` doesn't strictly implement `Sync`. This is still satisfactory as any user who needs `Sync` behavior can very easily wrap the `AppHandle` in a `Mutex` or other synchronization primitive to make it `Sync`.

Adding `AppHandle`, the user always has a simple option for accessing the main thread while the `Application`'s run loop owns it. `AppHandle` is only invalid before and after `Application::run`, and in both of those cases the user has control of the main thread. When Glazier has control of the main thread during `Application::run`, `AppHandle` is valid.

### CODE REVIEW

**Comment by jneem** [comment]
This makes sense to me! Actually, is there a reason to have the IdleHandle tied to the window at all? Why not just have it available from the application?

And I realize that this is sort of orthogonal to the point you're making, but often the easiest way to run something on the main thread is just to do it synchronously in one of the `WinHandler` methods -- those get called on the main thread.

**Comment by clavin** [comment]
`IdleHandle` exposes `WindowHandle` (and thus is tied to the concept of windows) in its API through its `add_idle` method. To reuse `IdleHandle` for `Application`, we would need to make `IdleHandle` generic, i.e. `IdleHandle<dyn WinHandler>` and `IdleHandle<dyn AppHandler>`. However, Glazier's API design seems to avoid generics from what I've seen, so the `AppHandle`/`IdleHandle` pair may be a better fit. I can think of some ways to make generic `IdleHandle<...>` more ergonomic (e.g. type aliases), but I digress.

In the pursuit of keeping the crate's API simple, I believe the `AppHandle`/`IdleHandle` duo would be better; though, I'm cool with whatever direction is decided upon here together 😄

I'm a bit worried about adding yet another "Handle" to Glazier's API, as folks might get confused when having to choose from the types `AppHandle`, `WindowHandle`, `IdleHandle`, and the closely-named `AppHandler` and `WinHandler` and when remembering their differences (the string "Handle" starts to lose a consistent meaning haha). As an aside, it might be worth cleaning up the App & Window nomenclature in Glazier's API in a separate PR since Glazier can still chose to make those decisions, if y'all are open to that! 🧹 🧼 🫧 I'm also open to other names for this type if y'all think there's a better choice 😊

**Comment by raphlinus** [comment]
The idea makes sense to me as well. To some extent, having things so intimately tied to windows is a holdover from Windows HWND, but a more correct model has basically a singleton application and then zero or windows.

**Comment by clavin** [comment]
Alright, I've completed and pushed implementations for macOS and Windows 😄 looking forward to opening this up for review feedback

File: src/common_util.rs [hunk 1]
```diff
@@ -50,14 +52,76 @@ pub fn strip_access_key(raw_menu_text: &str) -> String {
...
 }
 
-/// A trait for implementing the boxed callback hack.
-pub(crate) trait IdleCallback: Send {
-    fn call(self: Box<Self>, a: &mut dyn WinHandler);
+pub(crate) type IdleCallback = Box<dyn for<'a> FnOnce(&'a mut dyn WinHandler) + Send>;
+
+/// A sharable queue. Similar to a `std::sync::mpsc` channel, this queue is implmented as two types:
+/// [SharedEnqueuer] and [SharedDequeuer]. The enqueuer can be cloned, while the dequeuer cannot.
+/// The major difference between this and a channel is that the queue can be checked for emptiness.
```
Comment (jneem) [inline src/common_util.rs:59]: It isn't really "emptiness," right? For one thing, the channel and the flag aren't protected by a common lock so the empty flag won't always be up-to-date. But also, if someone just `try_dequeue`ed the last message then the queue will be empty but the empty flag won't say so.

Two possible suggestions:
1) rename `is_empty` to something like `needs_wake`, which I think is closer to the intended semantics, or
2) nudge the user towards the correct usage by deleting `is_empty` and replacing `enqueue` with
```
#[must_use]
fn enqueue(&self, t: T) -> NeedsWake { ... }

enum NeedsWake {
  Yes,
  No,
}
```

**Review by jneem — COMMENTED** [review]
Looks nice! Do you want to mark it as non-draft? I don't think the linux implementations are a prerequisite for merging.

I'm also not particularly qualified to review the windows and mac impls, but I'm inclined towards optimistic merging anyway :)

File: src/common_util.rs (hunk 1 above)
Comment (clavin) [reply src/common_util.rs:59]: That API is closer to what I started out with, but I couldn't figure out an apt name for the return type of `enqueue` 😅 thanks for reading my mind on this one! I'll get this swapped out with a single `enqueue` method.

**Comment by clavin** [comment]
I had some small things to clean up and polish, but I'm happy to open this for review now. 😊 And I appreciate the CI run! I have to rely on the CI output for the various Linux builds.

Some things that I wanted to point out to consider during review:

* While implementing the shared queue construct, I learned that `std::sync::mpsc`'s `Sender` and `Receiver` types are both `Send`, but `!Sync`. This makes sense to me, though this affects the original API I had proposed as `AppHandle` is no longer `Sync` (but is still `Send`!)
  * I could make `AppHandle` impl `Sync` by wrapping the `Sender` in a `Mutex` or `RwLock`, but I found this to be a bit overkill and rather leave this to the user if they _really_ need `Sync`.
* I tried to clone the existing implementation of `IdleHandle` where possible, so most usage of `unsafe` should have an equivalent window counterpart, with a few exceptions. Hopefully this makes those instances of unsafe code more trustworthy.

Thanks for taking the time to help me implement this! 😊

File: src/backend/windows/application.rs [hunk 2]
```diff
@@ -51,17 +54,21 @@ pub(crate) struct Application {
 struct State {
     quitting: bool,
     windows: HashSet<HWND>,
+    main_thread_cb_queue: (SharedEnqueuer<MainThreadCb>, SharedDequeuer<MainThreadCb>),
 }
 
 /// Used to ensure the window class is registered only once per process.
 static WINDOW_CLASS_REGISTERED: AtomicBool = AtomicBool::new(false);
 
+static MSG_RUN_MAIN_CB_QUEUE: AtomicU32 = AtomicU32::new(0);
```
Comment (clavin) [inline src/backend/windows/application.rs:63]: I think there's a better way to write this type with the recently-stabilized-but-not-yet-released-as-stable `std::sync::OnceCell` and, later, its lazy init equivalent. But it is not the future yet. 🤷

Comment (jneem) [reply src/backend/windows/application.rs:63]: It looks like we already depend on the `once_cell` crate, if you want to use that in the meantime.

**Comment by clavin** [comment]
(Pushed an update to the Windows impl to use the `once_cell`, but I need to test it when I get back to my Windows machine tonight)

File: src/common_util.rs [hunk 3]
```diff
@@ -50,14 +52,103 @@ pub fn strip_access_key(raw_menu_text: &str) -> String {
...
+    sender: mpsc::Sender<T>,
+    empty_flag: Arc<Mutex<bool>>,
 }
 
-impl<F: FnOnce(&mut dyn WinHandler) + Send> IdleCallback for F {
-    fn call(self: Box<F>, a: &mut dyn WinHandler) {
-        (*self)(a)
+impl<T> SharedEnqueuer<T> {
+    /// Adds a value to the queue. Returns `true` if the queue was empty before the value was added.
+    /// This should be used to wake the dequeuer.
```
Comment (jneem) [inline src/common_util.rs:98]:
```suggestion
    /// In this case, you need to wake the dequeuer.
```

**Review by jneem — APPROVED** [review]
Just one more `cargo fmt`, it looks like.

**Pull request merged**